zio-config/src/main/scala/zio/config/actions/ConfigDocs.scala

```diff
@@ -1,85 +1,93 @@
-            configValue.map(t => p.write(t)),
-            descAcc
+            path,
+            Descriptions(source.sourceDescription ++ descAcc.descriptions)
           )
         case ConfigDescriptor.Default(c, _) =>
-          loop(descAcc, c, docs, configValue, paths)
+          loop(descAcc, c, docs)
 
         case ConfigDescriptor.Describe(c, description) =>
-          loop(description :: descAcc, c, docs, configValue, paths)
+          loop(Descriptions(description :: descAcc.descriptions), c, docs)
 
         case ConfigDescriptor.Optional(c) =>
-          configValue match {
-            case Some(result) =>
-              result.fold(loop(descAcc, c, docs, None, paths))(v => loop(descAcc, c, docs, Some(v), paths))
-            case None =>
-              loop(descAcc, c, docs, None, paths)
-          }
+          loop(descAcc, c, docs)
 
-        case ConfigDescriptor.Nested(c, path) =>
-          loop(descAcc, c, docs, configValue, paths :+ path)
+        // intentional duplication of pattern matching to get over the type issues.
+        case a @ ConfigDescriptor.Nested(c, path) =>
+          a match {
+            case ConfigDescriptor.Nested(c, path) => ConfigDocs.NestedConfig(path, loop(descAcc, c, docs))
+          }
 
         case ConfigDescriptor.XmapEither(c, _, to) =>
-          configValue match {
-            case Some(v) =>
-              to(v).fold(_ => loop(descAcc, c, docs, None, paths), vv => loop(descAcc, c, docs, Some(vv), paths))
-            case None =>
-              loop(descAcc, c, docs, None, paths)
-          }
+          loop(descAcc, c, docs)
 
         case ConfigDescriptor.Zip(left, right) =>
-          configValue match {
-            case Some(tuple) =>
-              ConfigDocs.And(
-                loop(descAcc, left, docs, Some(tuple._1), paths),
-                loop(descAcc, right, docs, Some(tuple._2), paths)
-              )
-
-            case None =>
-              ConfigDocs.And(loop(descAcc, left, docs, None, paths), loop(descAcc, right, docs, None, paths))
-          }
+          ConfigDocs.And(
+            loop(descAcc, left, docs),
+            loop(descAcc, right, docs)
+          )
 
         case ConfigDescriptor.OrElseEither(left, right) =>
-          configValue match {
-            case Some(res) =>
-              res match {
-                case Left(vv) =>
-                  ConfigDocs.Or(loop(descAcc, left, docs, Some(vv), paths), loop(descAcc, right, docs, None, paths))
-
-                case Right(vv) =>
-                  ConfigDocs.Or(loop(descAcc, left, docs, None, paths), loop(descAcc, right, docs, Some(vv), paths))
+          ConfigDocs.Or(
+            loop(descAcc, left, docs),
+            loop(descAcc, right, docs)
+          )
+      }
+
+    loop(Descriptions(Nil), config, Empty())
+  }
+
+  def createDocWithValues[K, V, A](config: ConfigDescriptor[K, V, A], value: A): Either[String, ConfigDocs[K, V]] = {
+    val propertyTree = Write.write(config, value)
```

Comment:
I think Either makes makes sense. Given some value A, it doesn't necessarily mean, it can purely covert those values to its representation in a sys env / properties.